runners/flink/runner/src/test/java/org/apache/beam/runners/flink/streaming/UnboundedSourceWrapperTest.java

```diff
@@ -0,0 +1,318 @@
+            KV<Integer, Integer>,
+            TestCountingSource.CounterMark>> sourceOperator = new StreamSource<>(flinkWrapper);
+
+    setupSourceOperator(sourceOperator);
+
+
+    try {
+      sourceOperator.run(checkpointLock,
+          new Output<StreamRecord<WindowedValue<KV<Integer, Integer>>>>() {
+            private int count = 0;
+
+            @Override
+            public void emitWatermark(Watermark watermark) {
+            }
+
+            @Override
+            public void collect(
+                StreamRecord<WindowedValue<KV<Integer,Integer>>> windowedValueStreamRecord) {
+
+              count++;
+              if (count >= NUM_ELEMENTS) {
+                throw new SuccessException();
+              }
+            }
+
+            @Override
+            public void close() {
+
+            }
+          });
+    } catch (SuccessException e) {
+      // success
+    } catch (Exception e) {
+      fail("We caught " + e);
+    }
+  }
+
+  /**
+   * Creates a {@link UnboundedSourceWrapper} that has multiple readers per source, since we
+   * specify a parallelism higher than 1 and at runtime tell the source that it has 1 parallel
+   * this means that one source will manage multiple readers.
+   */
+  @Test
+  public void testWithMultipleReaders() throws Exception {
+    final int NUM_ELEMENTS = 20;
+    final Object checkpointLock = new Object();
+    PipelineOptions options = PipelineOptionsFactory.create();
+
+    // this source will emit exactly NUM_ELEMENTS across all parallel readers,
+    // afterwards it will stall. We check whether we also receive NUM_ELEMENTS
+    // elements later.
+    TestCountingSource source = new TestCountingSource(NUM_ELEMENTS);
+    UnboundedSourceWrapper<KV<Integer, Integer>, TestCountingSource.CounterMark> flinkWrapper =
+        new UnboundedSourceWrapper<>(options, source, 4);
+
+    assertEquals(4, flinkWrapper.getSplitSources().size());
+
+    StreamSource<WindowedValue<
+        KV<Integer, Integer>>,
+        UnboundedSourceWrapper<
+            KV<Integer, Integer>,
+            TestCountingSource.CounterMark>> sourceOperator = new StreamSource<>(flinkWrapper);
+
+    setupSourceOperator(sourceOperator);
+
+
+    try {
+      sourceOperator.run(checkpointLock,
+          new Output<StreamRecord<WindowedValue<KV<Integer, Integer>>>>() {
+            private int count = 0;
+
+            @Override
+            public void emitWatermark(Watermark watermark) {
+            }
+
+            @Override
+            public void collect(
+                StreamRecord<WindowedValue<KV<Integer,Integer>>> windowedValueStreamRecord) {
+
+              count++;
+              if (count >= NUM_ELEMENTS) {
+                throw new SuccessException();
+              }
+            }
+
+            @Override
+            public void close() {
+
+            }
+          });
+    } catch (SuccessException e) {
+      // success
+    } catch (Exception e) {
+      fail("We caught " + e);
```

Comment:
These blocks should signal to the outside of the try-catch that it was caught, and the test should fail if it was not (the test has terminated without consuming all of the expected available input).

generally, if the success is the last block before test completion,
```
try {...} catch (SuccessException e) {
  return;
}
fail("Read terminated without producing expected number of outputs")
```
or another appropriate message is sufficient; the resume-from-checkpoint needs a bit of additional work around it